### DISCUSSION TITLE
HsMod Web Api

### DISCUSSION THREAD

**Discussion by Pik-4** [Pik-4/HsMod#122] [Announcements]
HsMod的默认解析`Hearthstone\BepInEx\HsMod`下所有文件；默认监听本地`0.0.0.0:58744`；按`F4`可以重启webserver。如果需要其他额外接口，请提出说明。

HsMod parses all files in `Hearthstone\BepInEx\HsMod` by default; listens to local `0.0.0.0:58744` by default; press `F4` to restart the webserver. If you need other additional interfaces, please provide instructions. 

建议开发前删除`BepInEx\HsMod.cfg`，使其重新生成。

It is recommended to delete `BepInEx\HsMod.cfg`  before development and regenerate it.

目前已完成以下接口：

The following interfaces have been completed:

1. 进程状态，炉石登陆状态

   Process status, Hearthstone login status

   ```
   Request_Path: /alive
   Method:GET
   
   Return: json // text/html; charset=UTF-8
   {
   "pid":int, // hearthstone pid
   "login":string   // heartstone login status LoginCompleted
   }
   
   Example:
   {"pid":32600,"login":"True"}
   {"pid":32600,"login":"False"}
   
   ```

2. HsMod配置，可以从这里解析HsMod当前配置

   HsMod configuration, you can parse the current configuration of HsMod from here

   ```
   Request_Path: /hsmod.cfg
   Method:GET
   
   Return: text/plain; charset=UTF-8
   
   Example:
   ## Key Describe
   # Setting type: Int32
   # Default value: -1
   # Acceptable value range: From -1 to 100
   Key = Value
   ```

3. 皮肤映射配置

   Skin Mapping Configuration

   ```
   Request_Path: /hsskins.cfg
   Method:GET
   
   Return: text/plain; charset=UTF-8
   ```

4. jquery.min.js

   ```
   Request_Path: /jquery.min.js
   Method:GET
   
   Return:  text/javascript; charset=UTF-8
   
   /*! jQuery v3.7.1 | (c) OpenJS Foundation and other contributors | jquery.org/license */
   ```

5. 原有webpage；其中配置管理会去访问`./BepInEx/HsMod/config/index.html`，如果有配置管理工具，建议优先放在这里。

   The original webpage; the configuration management will access `./BepInEx/HsMod/config/index.html`. If there is a configuration management tool, it is recommended to put it here first.

   ```
   Request_Path: /collection
   Request_Path: /info
   Request_Path: /pack
   Request_Path: /lettuce
   Request_Path: /skins
   Request_Path: /mercenaries
   Request_Path: /matchlog
   Request_Path: /home
   Request_Path: /
   Method:GET
   
   ```

6. shell；需要开启webshell功能，shell开关选项不会被web api修改。

   shell; the `isWebshellEnable`  needs to be enabled, and the shell switch option will not be modified by the web api.

   ```
   Request_Path: /webshell
   Method:POST
   {'command':string}
   
   POST_Example:  //json
   {'command':'dir'}
   
   Return:"text/plain; charset=UTF-8"
   ```

7. 日志或程序信息

   Log or program information

   ```
   Request_Path: /bepinex.log  // bepinex log
   Request_Path: /bepinex.min.log  // bepinex log the last 666 lines
   Request_Path: /skins.log    // need login True; skin info
   Request_Path: /match.log    // may not exist
   Request_Path: /mercenaries.log  // need login True, mercenaries lettuce info
   
   Method:GET
   
   Return: text/plain; charset=UTF-8
   ```

8. 修改配置，需要两个参数，key和value，其中key对应配置名称（参考`GET /hsmod.cfg`）或代码中原始名称（`https://github.com/Pik-4/HsMod/blob/bepinex5/HsMod/Languages/zhCN.json`），如果修改成功，会返回json字符串，其中，成功时status为200，value为修改后的值。注意，如果传入错误参数，返回可能不为json；返回200时，如果value超出key的许可范围，会将key重置为默认值，并返回。

   

   To modify the configuration, two parameters are required, `key` and `value`, where `key` corresponds to the configuration name (refer to `GET /hsmod.cfg`) or the original name in the code(`https://github.com/Pik-4/HsMod/blob/bepinex5/HsMod/Languages/enUS.json`). If the modification is successful, a json string will be returned, where `status` is 200 when successful and `value` is the modified value. Note that if the wrong parameters are passed in, the return may not be json; when `200` is returned, if the value exceeds the permitted range of the `key`, the `key` will be reset to the default value and returned.

   ```
   Request_Path: /config
   Method:POST
   {'key':string,'value':string} // input
   {"status":int,"output":string}  // output
   
   POST_Example:  //json
   {'key':'表情数量','value':'0'}
   
   {'key':'receiveEnemyEmoteLimit.name','value':'0'}
   
   {'key':'receiveEnemyEmoteLimit.name','value':'-999'}
   
   {'key':'receiveEnemyEmoteLimit.name','value':'500'}
   
   Return:"text/plain; charset=UTF-8"
   
   {"status":200,"output":"0"}
   
   {"status":200,"output":"0"}
   
   {"status":200,"output":"-1"}
   
   {"status":200,"output":"100"}
   ```

9. 修改文件内容，需要两个参数，key和value，其中key对应配置文件名称，目前仅支持`HsSkins.cfg`的修改，如果修改成功，会返回json字符串，其中，成功时status为200，value为修改后的值。注意，如果传入错误参数，返回可能不为json。

   

   To modify the file content, two parameters are required, key and value. The key corresponds to the configuration file name. Currently, only `HsSkins.cfg` can be modified. If the modification is successful, a json string will be returned. If successful, the status is 200 and the value is the modified value. Note that if the wrong parameters are passed in, the return value may not be json.

   ```
   Request_Path: /update
   Method:POST
   {'key':string,'value':string} // input
   {"status":int,"output":string}  // output
   
   POST_Example:  //json
   {'key':'HsSkins.cfg','value':"50484:83406\n103767:78180,73709,83406,80187"}
   ```

**Comment by Lofuuzi** [comment]
感謝您提供的專案和思路！
我注意到並跟進上游專案庫進行的更新和相關變動，也將fork隱藏released。
不清楚您對此專案目前是否有未來的計劃，但尊重您的選擇並期待再見您回歸的那天🙏。
